### PR TITLE
fix(trashbin): don't overwrite existing file when restoring to a target (#35974)

### LINKED ISSUE #35974: Restore file from trashbin to a different place overwriting a file
### Steps to reproduce
```
  Scenario Outline: restoring a file to an already existing path overrides the file
    Given using <dav-path> DAV path
    And user "user0" has been created with default attributes and skeleton files
    And user "user0" has uploaded file with content "file to delete" to "/textfile0.txt"
    And user "user0" has uploaded file with content "PARENT textfile0 content" to "/PARENT/textfile0.txt"
    And user "user0" has deleted file "/textfile0.txt"
    When user "user0" restores the file with original path "/textfile0.txt" to "/PARENT/textfile0.txt" using the trashbin API
    Then the HTTP status code should be "204"
    # Sometimes "/PARENT/textfile0.txt" is found in the trashbin. Should it? Or not?
    # That seems to be what happens when the restore-overwrite happens properly,
    # The original /PARENT/textfile0.txt seems to be "deleted" and so goes to the trashbin
    #And as "user0" the file with original path "/PARENT/textfile0.txt" should not exist in trash
    And as "user0" file "/PARENT/textfile0.txt" should exist
    # sometimes the restore from trashbin does overwrite the existing file, but sometimes it does not. That is also surprising.
    # the current observed behavior is that if the original /PARENT/textfile0.txt ended up in the trashbin,
    # then the new /PARENT/textfile0.txt has the "file to delete" content.
    # otherwise /PARENT/textfile0.txt has its old content
    And the content of file "/PARENT/textfile0.txt" for user "user0" if the file is also in the trashbin should be "file to delete" otherwise "PARENT textfile0 content"
    #And the content of file "/PARENT/textfile0.txt" for user "user0" should be "file to delete"
    Examples:
      | dav-path |
      | old      |
      | new      |
```

### Expected behaviour
Maybe:
1) the restored file should just overwrite the existing file, and the existing file is gone (or is saved as an old version if `files_versions` is enabled), or;
2) The restored file has `(2)` on the end of its name, and the existing file stays, or;
3) ???

### Actual behaviour
Sometimes the restore from trashbin overwrites the existing file. When that happens, the existing file is found in the trashbin. I guess that the code-path deletes the existing file (triggering an event that causes it to be saved into the trashbin). And the restored file has the correctly-restored content.

But also sometimes the existing file is not found in the trashbin, and has not been overwritten (it still has the old content). But the API call to restore from trashbin still returns `204` success.

### Server configuration
Local runs of the above scenario in a development environment on top of current `master`
And in drone CI it happens also.

### PR DESCRIPTION
**Pull request by DeepDiver1975**
## Summary

Fixes #35974 — restoring a file from the trashbin to an **explicit target location** could silently **overwrite an existing file there, causing data loss**.

`Trashbin::restore($filename, $targetLocation = null)` only resolves a collision-safe unique name on the default path (`$targetLocation === null`, restore to original location). When an explicit `$targetLocation` is supplied — the WebDAV *"restore to a different place"* path (`apps/dav/lib/TrashBin/…`) — it skipped the unique-name step and called `View::rename($source, $target)` directly, overwriting whatever already existed at the target.

## Change

Add a data-loss guard just before the `rename`: if an **explicit** target already exists, refuse the restore and return `false` instead of clobbering it.

```php
if ($view->file_exists($target)) {
    return false;
}
```

- The bool return contract is preserved and propagates cleanly through `TrashBinManager::restore()` → `AbstractTrashBinNode::restore()`.
- The default null-target path already resolves to a unique, non-existing name via `getUniqueFilename()`, so **normal restores are unaffected** (the guard is never hit there).
- Failing (rather than auto-renaming) is the conservative choice for the explicit-target case: the caller asked for that exact path, so a clear failure is safer and more predictable than silently restoring to a different name.

## Tests

Adds `TrashbinTest::testRestoreToExplicitTargetDoesNotOverwriteExisting()`: delete a file into the trashbin, create a *different* existing file at the explicit target, attempt the restore, and assert (a) it returns `false` and (b) the existing target file's content is unchanged.

> Tagged `@group DB` (existing class). `php -l` clean; full PHPUnit not run in the preparation environment.

> Note: `owncloud/core` is in maintenance mode; targets installations on classic ownCloud 10.x.

---

🤖 This PR was prepared by the Claude Code review agent from the analysis of #35974. Please review carefully before merging.

### CODE REVIEW

**Review by DeepDiver1975 — COMMENTED** [review]
🤖 Automated code review by Claude Code review agent

## Overview

This PR adds a data-loss guard in `Trashbin::restore()`: when an explicit `$targetLocation` is supplied and a file already exists at the resolved target, it returns `false` instead of calling `View::rename()` and clobbering the file. A regression unit test is added. The change is small, well-commented, and the intent (preventing data loss) is excellent.

I traced the change against the actual code paths and the original issue #35974. Unfortunately I found a **blocking correctness problem**: the guard does not fix the bug as it actually occurs over WebDAV.

## Correctness

**Placement (good):** The guard sits correctly *after* the source-existence check and *before* `$view->rename($source, $target)` (Trashbin.php ~L584–591). Returning `false` is consistent with the method's documented `bool` contract.

**Null-target path (good, unaffected):** When `$targetLocation === null`, the target is resolved via `getUniqueFilename()` to a guaranteed non-existing name, so `file_exists($target)` is always `false` and the guard never blocks a normal restore. Confirmed.

**BLOCKING — the guard never fires on the real WebDAV "restore to a different place" path.** The explicit-target caller is the WebDAV MOVE flow: `Directory::moveInto()` → `AbstractTrashBinNode::restore()` → `TrashBinManager::restore()` → `Trashbin::restore($path, $targetLocation)`. But before `Trashbin::restore()` is ever reached, Sabre's `CorePlugin::httpMove()` has already handled the destination:

```php
// lib/composer/sabre/dav/lib/DAV/CorePlugin.php  (httpMove)
if ($moveInfo['destinationExists']) {
    $this->server->tree->delete($moveInfo['destination']);   // <-- existing target deleted FIRST
    $this->server->emit('afterUnbind', [$moveInfo['destination']]);
}
$this->server->tree->move($path, $moveInfo['destination']);   // <-- only now does restore() run
```

`Overwrite` defaults to `T` when the header is absent (`Server::getCopyAndMoveInfo()`), so `destinationExists` is `true` and Sabre **pre-deletes the existing target** before the move. By the time our guard runs, `$view->file_exists($target)` is already `false`, so the guard is a no-op and the restore proceeds — exactly the data loss we meant to prevent.

This matches the original issue #35974 verbatim, which reports: *"the code-path deletes the existing file (triggering an event that causes it to be saved into the trashbin)."* That is Sabre's `tree->delete($destination)`. The reporter also notes the behavior is intermittent — the guard addresses none of that.

With `Overwrite: F`, `getCopyAndMoveInfo()` throws `PreconditionFailed` (412) before restore runs, so the guard is moot there too.

Net: the only path that reaches the guard with a pre-existing target is a **direct PHP call** to `Trashbin::restore($filename, $explicitTarget)` that bypasses Sabre's pre-delete — which is precisely (and only) what the new unit test does. The fix passes its own test but does not fix the reported WebDAV bug.

If a fix at this layer is desired, it likely belongs in the DAV layer — e.g. `Directory::moveInto()` should reject the restore when the destination already exists *before* Sabre's pre-delete (or `httpMove`/the trashbin restore endpoint must opt out of the destination pre-delete). A guard inside `Trashbin::restore()` is too late in the chain.

## Tests

The test `testRestoreToExplicitTargetDoesNotOverwriteExisting()` is internally well-formed and would **pass in CI**: it follows the established pattern of the neighbouring `testRestoreFileDoesNotOverwriteExistingInRoot()`, uses the correct trash filename format `file1.txt.d<mtime>` via `$trashedFile->getMtime()`, uses real helpers (`Helper::getTrashFiles`, user folder nodes), and the assertions (`assertFalse(restore(...))` + unchanged `existing.txt` content) are valid for the direct-call path.

However the test gives **false confidence**: it exercises a direct `Trashbin::restore()` call, not the WebDAV MOVE flow where the bug lives. It therefore green-lights a change that does not fix #35974 in production. A meaningful regression test would drive the actual DAV trashbin restore endpoint (as the issue's Gherkin scenario does) and assert the destination is untouched and a 4xx/no-overwrite results.

## Edge cases

- Direct `Trashbin::restore()` callers with an explicit target ARE now protected (positive, though no such caller exists in core today besides tests/3rd-party).
- The `ajax/undelete.php` path only ever passes a null target, so it is unaffected (correct).
- Folder restores via WebDAV hit the same Sabre pre-delete behavior, so they are equally unprotected.
- No new path is broken by the change; the risk is purely that it under-delivers, not that it regresses.

## Verdict

**changes-requested (blocking).** The fix is safe and does not regress anything, but it does not actually fix issue #35974: on the real WebDAV restore-to-target path, Sabre deletes the pre-existing destination *before* `Trashbin::restore()` runs, so the new guard never fires. The accompanying test passes only because it bypasses that flow via a direct PHP call, giving false confidence. Please move the protection to the DAV layer (reject/short-circuit the restore when the destination exists, before Sabre's `httpMove` pre-delete) and add a test that drives the WebDAV MOVE/restore endpoint end-to-end.

**Review by dj4oC — CHANGES_REQUESTED** [review]
Adds a `$view->file_exists($target)` guard in `Trashbin::restore()` intended to stop the WebDAV restore-to-target flow from silently overwriting an existing file (issue #35974).

### Findings
- **Security/Stability (blocking, data-loss risk) — the fix does not address the reported vulnerability on the real path.** Two independent review passes traced the actual WebDAV MOVE call chain (`CorePlugin::httpMove()` → Sabre `Tree::move()` → `Directory::moveInto()` → `AbstractTrashBinNode::restore()` → `Trashbin::restore()`) and confirmed: when the destination already exists and the client sends the RFC4918-default `Overwrite: T` (or omits the header, which is what this repo's own acceptance-test harness does), **Sabre's core plugin deletes the destination itself before `moveInto()`/`restore()` ever runs** — so `$view->file_exists($target)` is already `false` by the time the new guard executes, making it dead code on the real path. If a client sends `Overwrite: F` instead, Sabre throws `412` before `restore()` is ever reached. Either way, the new guard only protects direct/programmatic callers that bypass Sabre entirely — exactly what this PR's own regression test does (`TrashbinTest::testRestoreToExplicitTargetDoesNotOverwriteExisting` calls `Trashbin::restore()` directly).
- Compounding this: there's an existing, unmodified acceptance scenario (`tests/acceptance/features/apiTrashbinRestore/trashbinRestoreOc10Issue35974.feature`) documenting the historical "restore to occupied target" behavior (expects success + old target trashed) that this PR neither updates nor reconciles with.
- Also flagged: `file_exists()` + rename is not atomic (CWE-367 TOCTOU) — a concurrent write to the target between check and move could still cause data loss even if the guard were reachable.
- Performance: no findings — a single stat call, not in a loop.
- Test coverage: the included unit test does not exercise the real WebDAV/Sabre-driven restore path, so it cannot catch the dead-code issue above.

### Verdict
Changes requested — this needs a fix (and test) that actually exercises the WebDAV MOVE endpoint end-to-end, and reconciliation with the existing `trashbinRestoreOc10Issue35974.feature` scenario, before this can be considered to fix #35974. Given the data-loss-risk nature of this code path, recommend explicit human maintainer sign-off in addition to a corrected implementation.

🤖 Automated review by Claude Code (security · stability · performance · coverage)

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by jvillafanez — DISMISSED** [review]

File: apps/dav/lib/TrashBin/TrashBinPlugin.php [hunk 1]
```diff
@@ -40,6 +41,41 @@ public function initialize(Server $server) {
...
+		if (!($sourceNode instanceof ITrashBinNode)) {
+			// not a restore - let the default MOVE handling proceed
+			return true;
+		}
+
+		if ($this->server->tree->nodeExists($destinationPath)) {
+			throw new Forbidden(
+				'Could not restore, target "' . $destinationPath . '" already exists'
+			);
+		}
```
Comment (jvillafanez) [inline apps/dav/lib/TrashBin/TrashBinPlugin.php:76]: I think we should check for the `Overwrite` header. It would be weird if the client sends a `Overwrite: T` and the server responds with an "already exists" error when it has been explicitly told to overwrite the file.
